transport.go

```diff
@@ -0,0 +1,119 @@
+package legs
+
+import (
+	"context"
+	"fmt"
+	"io/ioutil"
+	"os"
+	"sync"
+
+	dt "github.com/filecoin-project/go-data-transfer"
+	datatransfer "github.com/filecoin-project/go-data-transfer/impl"
+	dtnetwork "github.com/filecoin-project/go-data-transfer/network"
+	gstransport "github.com/filecoin-project/go-data-transfer/transport/graphsync"
+	"github.com/ipfs/go-datastore"
+	"github.com/ipfs/go-graphsync"
+	gsimpl "github.com/ipfs/go-graphsync/impl"
+	gsnet "github.com/ipfs/go-graphsync/network"
+	"github.com/ipld/go-ipld-prime"
+	"github.com/libp2p/go-libp2p-core/host"
+	pubsub "github.com/libp2p/go-libp2p-pubsub"
+	pubsubpb "github.com/libp2p/go-libp2p-pubsub/pb"
+	"github.com/minio/blake2b-simd"
+)
+
+const (
+	// directConnectTicks makes pubsub check it's connected to direct peers every N seconds.
+	directConnectTicks uint64 = 30
+)
+
+// LegTransport wraps all the assets to set-up the data transfer.
+type LegTransport struct {
+	tmpDir string
+	t      dt.Manager
+	Gs     graphsync.GraphExchange
+	topic  *pubsub.Topic
+
+	lk   sync.Mutex
+	refc int
```

Comment:
consider updates to `refc` via [atomic](https://pkg.go.dev/sync/atomic) rather than protecting it with a mutex?